### PR TITLE
[IntelliJ Plugin] Add code folding support

### LINKED ISSUE #12759: Enable code folding capability in ballerina IntelliJ plugin
**Description:**
Implementation for automatic code folding ranges detection has been disabled in IntelliJ IDEA plugin. Therefore need to revisit the implementation and enable code folding capability.

**Steps to reproduce:**

**Affected Versions:**
0.990.0 and below.

**OS, DB, other environment details and versions:**

**Related Issues (optional):**

**Suggested Labels (optional):**

**Suggested Assignees (optional):**

### PR DESCRIPTION
**Pull request by NipunaRanasinghe**
## Purpose
This PR adds code folding region detection in IDEA plugin for,
- Objects 
- Records
- Functions
- Services
- Workers
- Documentation
- Annotations
- Multi-line comments

Fixes #12759.

### CODE REVIEW

File: tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/editor/BallerinaFoldingBuilder.java [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+    }
+
+    private void buildMultiCommentFoldingRegions(@NotNull List<FoldingDescriptor> descriptors,
+            @NotNull PsiElement root) {
+
+        Collection<PsiComment> comments = PsiTreeUtil.findChildrenOfType(root, PsiComment.class);
+
+        for (PsiComment comment : comments) {
+            PsiElement prevSibling = getPreviousElement(comment);
+            if (!(prevSibling instanceof PsiComment)) {
```
Comment (Shan1024) [inline tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/editor/BallerinaFoldingBuilder.java]: Invert the `if` logic here to remove the large nested block.

File: tool-plugins/intellij/src/test/resources/testData/folding/Annotation.bal [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+service helloWorld on helloWorldEP <fold text='{...}'>{
+
+    @http:ResourceConfig <fold text='{...}'>{
+        methods: ["GET"],
+        path: "/"
+    }</fold>
+    resource function sayHello(http:Caller caller, http:Request req) <fold text='{...}'>{
+
+    }</fold>
+}</fold>
```
Comment (Shan1024) [inline tool-plugins/intellij/src/test/resources/testData/folding/Annotation.bal]: Missing newline.

File: tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/editor/BallerinaFoldingBuilder.java [hunk 3]
```diff
@@ -0,0 +1,300 @@
+/*
+ *  Copyright (c) 2018, WSO2 Inc. (http://www.wso2.org) All Rights Reserved.
```
Comment (Shan1024) [inline tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/editor/BallerinaFoldingBuilder.java]: Update the year to 2019.

Comment (NipunaRanasinghe) [reply tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/editor/BallerinaFoldingBuilder.java]: Updated

File: tool-plugins/intellij/src/test/resources/testData/folding/Annotation.bal (hunk 2 above)
Comment (NipunaRanasinghe) [reply tool-plugins/intellij/src/test/resources/testData/folding/Annotation.bal]: Fixed.

File: tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/editor/BallerinaFoldingBuilder.java (hunk 1 above)
Comment (NipunaRanasinghe) [reply tool-plugins/intellij/src/main/java/io/ballerina/plugins/idea/editor/BallerinaFoldingBuilder.java]: Changed.

**Pull request merged**